src/main/java/org/jitsi/videobridge/AbstractEndpointMessageTransport.java

```diff
@@ -0,0 +1,436 @@
+        Object src,
+        JSONObject jsonObject)
+    {
+        String to = (String)jsonObject.get("to");
+
+        // First insert the "from" to prevent spoofing.
+        jsonObject.put("from", getId(jsonObject.get("from")));
+        Conference conference = getConference();
+        if (conference == null || conference.isExpired())
+        {
+            logger.warn(
+                "Unable to send EndpointMessage, conference is null or expired");
+            return;
+        }
+
+        List<AbstractEndpoint> endpointSubset;
+        if ("".equals(to))
+        {
+            // Broadcast message
+            endpointSubset = new LinkedList<>(conference.getEndpoints());
+            endpointSubset.removeIf(
+                e -> e.getID().equalsIgnoreCase(getId()));
+        }
+        else
+        {
+            // 1:1 message
+            AbstractEndpoint targetEndpoint = conference.getEndpoint(to);
+            if (targetEndpoint != null)
+            {
+                endpointSubset = Collections.singletonList(targetEndpoint);
+            }
+            else
+            {
+                endpointSubset = Collections.emptyList();
+                logger.warn(
+                    "Unable to find endpoint " + to
+                        + " to send EndpointMessage");
+            }
+        }
+
+        sendMessageToEndpoints(jsonObject.toString(), endpointSubset);
+    }
+
+    /**
+     * Sends a specific message coming from this endpoint to other endpoints in
+     * the conference.
+     * @param msg the message to send.
+     * @param endpoints the list of endpoints to receive the message.
+     */
+    protected void sendMessageToEndpoints(
+        String msg,
+        List<AbstractEndpoint> endpoints)
+    {
+        Conference conference = getConference();
+        if (conference != null)
+        {
+            conference.sendMessage(msg, endpoints, true);
+        }
+    }
+
+    /**
+     * @return the associated {@link Conference} or {@code null}.
+     */
+    protected Conference getConference()
+    {
+        return endpoint != null ? endpoint.getConference() : null;
+    }
+
+    /**
+     * @return the ID of the associated endpoint or {@code null}.
+     */
+    private String getId()
+    {
+        return getId(null);
+    }
+
+    /**
+     * @return the ID of the associated endpoint or {@code null}.
+     * @param id a suggested ID.
+     */
+    protected String getId(Object id)
+    {
+        return endpoint != null ? endpoint.getID() : null;
+    }
+
+    /**
+     * Notifies this {@code Endpoint} that a {@code PinnedEndpointChangedEvent}
+     * has been received.
+     *
+     * @param src the transport channel by which {@code jsonObject} has
+     * been received.
+     * @param jsonObject the JSON object with {@link Videobridge#COLIBRI_CLASS}
+     * {@code PinnedEndpointChangedEvent} which has been received.
+     */
+    abstract protected void onPinnedEndpointChangedEvent(
```

Comment:
i noticed a method above we just had a no-op implementation instead of making it abstract, is there a reason we force this to be implemented but `notifyTransportChannelConnected` has a default no-op?

Reply:
or maybe it's better to ask, why do some of the message handlers below have default implementations and some don't?

Reply:
If I could use a single implementation I just placed it in the base class. If I ended up with one actual implementation and a no-op, I added the no-op to the base class and overrode it. If I had two different implementations I made the base method abstract.

Many of these fell to the third category only because I wanted to log a warning in case of Octo (I don't expect them to go through Octo, and I'd like to know if something happens and they end up going). I could move these to the base implementation if you this that's a better approach.